include/seqan3/range/view/pairwise_combine.hpp

```diff
@@ -0,0 +1,595 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::view::pairwise_combine.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <cmath>
+
+#include <seqan3/range/concept.hpp>
+#include <seqan3/range/view/detail.hpp>
+#include <seqan3/std/ranges>
+#include <seqan3/std/view/view_all.hpp>
+
+namespace seqan3::detail
+{
+/*!\brief Generates all pairwise combinations of the elements in the source range.
+ * \ingroup view
+ * \tparam source_range_type The type of the source range; must model std::ranges::View, std::ranges::ForwardRange
+ *
+ * \details
+ *
+ * This view will provide a convenient way to iterate over all pairwise combinations of a given source range
+ * while the order of selection does not matter. A source range with `n` elements will therefore have `n choose 2`
+ * possible combinations.
+ *
+ * ### Example
+ *
+ * \include test/snippet/range/view/pairwise_combine.cpp
+ */
+template <std::ranges::View source_range_type>
+//!\cond
+    requires  std::ranges::ForwardRange<source_range_type>
```

Comment:
Right. Actually I had this before but then dropped it because I thought this restriction wouldn't be necessary. But I also did not implement it right then. But after doing some more reading on the topic I came to the conclusion that this requirement does not really end up to be a restriction, given that only non-common bidirectional ranges would be downgraded to forward ranges if converted to a common range, but need a linear construction anyway. Just they can't be decremented anymore. I can live with that.

Reply:
Having interdependencies between the different range concepts is difficult to understand for the user and error-prone to program. I agree though that it is often easier to do and our focus should be in getting the features out.
Do I understand correctly that currently preserving size and anything above forward-range requires CommonRange? Then I would propose one of the following:
1. clearly documenting this in the table by writing *preserved iff CommonRange<urng_t>* into the column; and adding tests that guarantee that the view still works for all types of input (i.e. properly downgrades a Bidirectional-but-not-Common-Range to Forward).
2. Just require CommonRange for everything. That will make it easier to test and document. People can still use `| view::common` at the cost of some performance...

Regarding the preservedness of SizedRange: 
I still don't think you can guarantee it, you can only preserve it in certain cituations. What are the rules for preserving? Is it only preserved for RandomAccess (
RandomAccessRange+CommonRange implies SizedRange so you don't need extra documentation for that). What about Bidirectional+Sized+Common?